### DISCUSSION TITLE
How to make RLS policy that depends on column value?

### DISCUSSION THREAD

**Discussion by esDotDev** [supabase/supabase#5033] [Questions] [answered]
I have a `keys` table, which allows users to unlock the pro features in my app.

So I would like to allow an anonymous client who has one of these magic keys, to check if it exists. Without allowing full read-access for every row to all anon users.

So, for example, these calls would work, _only_ if `ABCDEFG` exists in the key table and only on that specific row:
`supabase.from('keys').select('id').eq('key', 'ABCDEFG');`
`supabase.from('keys').update({'hasBeenUsed', 'true'}).eq('key', 'ABCDEFG');`

This would not work, as user can not read the entire table, otherwise they could just select all the auth keys in the table:
`supabase.from('keys').select('id');`

**Comment by GaryAustin1** [comment, marked as answer]
You might want to lock down select with RLS and then use a rpc call to a postgres function with the key as a parameter.  This function would then do a select on just that key.  The function would also have to have "security definer" in its description to bypass RLS.
RLS for updates/inserts/deletes can look at incoming column data, but select does not get sent column data obviously.

**Reply by GaryAustin1** [reply]
This is an example of the settings in the function UI to fetch a record from the table with a text parameter as the key and security definer set.
I don't know your skill level, but there are lots of examples looking up postgres functions using the sql editor with create function syntax where you role your own function without the Supabase function UI, including in the Supabase docs.
In my code room_joins has a text key room_key.

![image](https://user-images.githubusercontent.com/54564956/149860738-62af4bf3-a2e3-47ab-ae6c-ec6996803887.png)

edit:   If that does not work, drop the begin and end in the code section and instead of plpgsql just select sql.   My function is actually more complicated so tried to simplify it quickly.  Plpgsql is a programming language in functions that lets you do if thens and other things, but not needed for just a select...

**Reply by esDotDev** [reply]
Awesome, this was super helpful thx. Unblocked me on a few fronts:
1. I didn't know that UI existed, and was trying to do it all by raw postgres
2. Now I understand what you meant by the security_definer option
3. begin/end seem to be crucial, or it won't let me save, so thx again!

**Reply by esDotDev** [reply]
In the end just had to do something like :
```
begin
return(select id from keys where key = value);
end
```
Then I can call with:
```
var response = await supabaseClient.rpc('checkKey', params: {'value': 'SOMEKEY'}).execute();
if (response.data != null) {
  print('Key is good!');
}
```
thx!

**Reply by esDotDev** [reply]
For completeness, I actually ended up getting a little more complex, the function now increments an internal 'useCount' and logs firstUseDate as well.
```
declare
  count int;
begin
SELECT "useCount" into count FROM keys WHERE key = value;
if count = 0 then
   UPDATE keys SET "firstUseDate" = NOW();
end if;
UPDATE keys SET "useCount" = count+1 WHERE key = value;
return(count);
end
```

**Comment by esDotDev** [comment]
Ah I see, thx. Do you know of any example snippets showing how I might create and test these RPCs? I'm having a tough time finding good code examples for postgres.
